## Assignment 1 - The Bean Project


In [1]:
import pandas as pd

from sklearn.preprocessing import StandardScaler
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

# Loading dataset 
data = pd.read_csv('Dry_Bean_Dataset.csv')

# Head dataset 
data.head()

# data['Class'].value_counts()

Area  Perimeter  MajorAxisLength  MinorAxisLength  AspectRation  \
0  28395    610.291       208.178117       173.888747      1.197191   
1  28734    638.018       200.524796       182.734419      1.097356   
2  29380    624.110       212.826130       175.931143      1.209713   
3  30008    645.884       210.557999       182.516516      1.153638   
4  30140    620.134       201.847882       190.279279      1.060798   

   Eccentricity  ConvexArea  EquivDiameter    Extent  Solidity  roundness  \
0      0.549812       28715     190.141097  0.763923  0.988856   0.958027   
1      0.411785       29172     191.272751  0.783968  0.984986   0.887034   
2      0.562727       29690     193.410904  0.778113  0.989559   0.947849   
3      0.498616       30724     195.467062  0.782681  0.976696   0.903936   
4      0.333680       30417     195.896503  0.773098  0.990893   0.984877   

   Compactness  ShapeFactor1  ShapeFactor2  ShapeFactor3  ShapeFactor4  Class  
0     0.913358      0.007332      0.003147      0.834222      0.998724  SEKER  
1     0.953861      0.006979      0.003564      0.909851      0.998430  SEKER  
2     0.908774      0.007244      0.003048      0.825871      0.999066  SEKER  
3     0.928329      0.007017      0.003215      0.861794      0.994199  SEKER  
4     0.970516      0.006697      0.003665      0.941900      0.999166  SEKER

# Task 1 

In [2]:
# # Finding Missing values
# missing_values = data.isnull().sum()

# # Drop missing rows values
# data = data.dropna()

X = data.drop('Class', axis=1)
y = data['Class']

In [3]:
# Satndard Scaler  
scaler = preprocessing.StandardScaler().fit(X)
X_scaled = scaler.transform(X)


In [4]:
# Here taking Training 80%, Testing 20%
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=30)


## Implementation classification models:
####  Accuracy, Precision, Recall and Confusion matrices

### 1. KNN 

In [5]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
knn_pred = knn.predict(X_test)

# K-Nearest Neighbour
knn_accuracy = accuracy_score(y_test, knn_pred)
knn_precision = precision_score(y_test, knn_pred, average='weighted')
knn_recall = recall_score(y_test, knn_pred, average='weighted')
knn_confusion_matrix = confusion_matrix(y_test, knn_pred)

# Results
print("--------K-Nearest Neighbour--------")
print("\nAccuracy =", knn_accuracy*100)
print("Precision =", knn_precision*100)
print("Recall =", knn_recall*100)
print("")
print("-----------------------------------")
print("Confusion Matrix:\n")
print(knn_confusion_matrix)
print("\n--------------------------------")
print("Classification Report:\n")
reportknn = classification_report(y_test, knn_pred)
print(reportknn)

--------K-Nearest Neighbour--------

Accuracy = 92.9122291590158
Precision = 92.92990003609243
Recall = 92.9122291590158

-----------------------------------
Confusion Matrix:

[[243   0  16   0   0   1   3]
 [  0  99   0   0   0   0   0]
 [  7   0 282   0   5   1   2]
 [  0   0   0 649   1   8  47]
 [  0   0   9   1 391   0   7]
 [  1   0   0   7   1 390   8]
 [  1   0   0  50   8   9 476]]

--------------------------------
Classification Report:

              precision    recall  f1-score   support

    BARBUNYA       0.96      0.92      0.94       263
      BOMBAY       1.00      1.00      1.00        99
        CALI       0.92      0.95      0.93       297
    DERMASON       0.92      0.92      0.92       705
       HOROZ       0.96      0.96      0.96       408
       SEKER       0.95      0.96      0.96       407
        SIRA       0.88      0.88      0.88       544

    accuracy                           0.93      2723
   macro avg       0.94      0.94      0.94      2723
weigh

### 2. DecisionTree


In [6]:
decision_tree = DecisionTreeClassifier(random_state=30)
decision_tree.fit(X_train, y_train)
dt_pred = decision_tree.predict(X_test)

# Decision Tree
dt_accuracy = accuracy_score(y_test, dt_pred)
dt_precision = precision_score(y_test, dt_pred, average='macro')
dt_recall = recall_score(y_test, dt_pred, average='macro')
dt_confusion_matrix = confusion_matrix(y_test, dt_pred)

# Results
print("----------Decision Tree---------")
print("\nAccuracy =", dt_accuracy*100)
print("Precision =", dt_precision*100)
print("Recall =", dt_recall*100)
print("")
print("--------------------------------")
print("Confusion Matrix:")
print(dt_confusion_matrix)
print("")
print("--------------------------------")
print("Classification Report:\n")
reportdt = classification_report(y_test, dt_pred)
print(reportdt)

----------Decision Tree---------

Accuracy = 89.46015424164524
Precision = 91.02835660836588
Recall = 91.13636244319144

--------------------------------
Confusion Matrix:
[[240   0  14   0   0   5   4]
 [  0  99   0   0   0   0   0]
 [ 21   0 268   0   6   1   1]
 [  0   0   0 625   9   7  64]
 [  2   0   7   4 380   0  15]
 [  7   0   1  15   0 375   9]
 [  6   0   2  64   9  14 449]]

--------------------------------
Classification Report:

              precision    recall  f1-score   support

    BARBUNYA       0.87      0.91      0.89       263
      BOMBAY       1.00      1.00      1.00        99
        CALI       0.92      0.90      0.91       297
    DERMASON       0.88      0.89      0.88       705
       HOROZ       0.94      0.93      0.94       408
       SEKER       0.93      0.92      0.93       407
        SIRA       0.83      0.83      0.83       544

    accuracy                           0.89      2723
   macro avg       0.91      0.91      0.91      2723
weighted a

### 3. Logistic Regression 

In [7]:
logistic_reg = LogisticRegression(solver='liblinear', max_iter=1000)
logistic_reg.fit(X_train, y_train)
lr_pred = logistic_reg.predict(X_test)

# Logistic Regression
lr_accuracy = accuracy_score(y_test, lr_pred)
lr_precision = precision_score(y_test, lr_pred, average='weighted')
lr_recall = recall_score(y_test, lr_pred, average='weighted')
lr_confusion_matrix = confusion_matrix(y_test, lr_pred)

# Results
print("----------Logistic Regression---------\n")
print("Accuracy =", lr_accuracy*100)
print("Precision =", lr_precision*100)
print("Recall =", lr_recall*100)
print("")
print("--------------------------------------")
print("Confusion Matrix:")
print(lr_confusion_matrix)
print("")
print("--------------------------------")
print("Classification Report:\n")
reportlr = classification_report(y_test, lr_pred)
print(reportlr)

----------Logistic Regression---------

Accuracy = 92.06757253029747
Precision = 92.1885282156184
Recall = 92.06757253029747

--------------------------------------
Confusion Matrix:
[[237   0  14   0   0   2  10]
 [  0  99   0   0   0   0   0]
 [  6   0 272   0  10   1   8]
 [  1   0   0 640   3  10  51]
 [  2   0   5   2 389   0  10]
 [  3   0   0   3   1 386  14]
 [  0   0   1  47   7   5 484]]

--------------------------------
Classification Report:

              precision    recall  f1-score   support

    BARBUNYA       0.95      0.90      0.93       263
      BOMBAY       1.00      1.00      1.00        99
        CALI       0.93      0.92      0.92       297
    DERMASON       0.92      0.91      0.92       705
       HOROZ       0.95      0.95      0.95       408
       SEKER       0.96      0.95      0.95       407
        SIRA       0.84      0.89      0.86       544

    accuracy                           0.92      2723
   macro avg       0.94      0.93      0.93      2723

### 4. Random Forest

In [8]:
random_forest = RandomForestClassifier(random_state=30)
random_forest.fit(X_train, y_train)
rf_pred = random_forest.predict(X_test)

rf_accuracy = accuracy_score(y_test, rf_pred)
rf_precision = precision_score(y_test, rf_pred, average='weighted')
rf_recall = recall_score(y_test, rf_pred, average='weighted')
rf_confusion_matrix = confusion_matrix(y_test, rf_pred)

# Print the results
print("------------Random Forest-------------")
print("Accuracy =", rf_accuracy*100)
print("Precision =", rf_precision*100)
print("Recall =", rf_recall*100)
print("")
print("--------------------------------------")
print("Confusion Matrix:")
print(rf_confusion_matrix)
print("")
print("--------------------------------")
print("Classification Report:\n")
reportrf = classification_report(y_test, rf_pred)
print(reportrf)

------------Random Forest-------------
Accuracy = 93.68343738523687
Precision = 93.67568656898777
Recall = 93.68343738523687

--------------------------------------
Confusion Matrix:
[[246   0  12   0   0   2   3]
 [  0  99   0   0   0   0   0]
 [ 11   0 277   0   6   1   2]
 [  0   0   0 667   2   6  30]
 [  1   0   6   2 393   0   6]
 [  1   0   0   8   0 394   4]
 [  0   0   1  50   8  10 475]]

--------------------------------
Classification Report:

              precision    recall  f1-score   support

    BARBUNYA       0.95      0.94      0.94       263
      BOMBAY       1.00      1.00      1.00        99
        CALI       0.94      0.93      0.93       297
    DERMASON       0.92      0.95      0.93       705
       HOROZ       0.96      0.96      0.96       408
       SEKER       0.95      0.97      0.96       407
        SIRA       0.91      0.87      0.89       544

    accuracy                           0.94      2723
   macro avg       0.95      0.95      0.95      2723

###
 In this above case, the Random Forest Model has the highest scores for Accuracy, Precision, and Recall. 
 1. Accuracy - measuring how accurately predictions are made overall.
 2. Precision - measurement for how well a model can predict the future.
 3. Recall - measures how well a model can find all instances of success. 
 

In this first task we have calculated the accuracy which is based on the confusion matrix which gives us the outcome of the results. 
 
Overall, all of the models have decent predictive potential, but in this case, Random Forest perform best.
 
In the above scenario we seen that all models shows their values, all models exhibited accuracy rates more than approximately 89%, which is a respectable outcome.

# Task 2 

In [9]:
class_counts = data['Class'].value_counts()
print(class_counts)

DERMASON    3546
SIRA        2636
SEKER       2027
HOROZ       1928
CALI        1630
BARBUNYA    1322
BOMBAY       522
Name: Class, dtype: int64


### Method 1 : Using SMOTE for Descision Tree 

In [10]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from imblearn.over_sampling import SMOTE
from sklearn.metrics import classification_report

X = data.drop(columns=['Class'])
y = data['Class']

scaler = preprocessing.StandardScaler().fit(X)
X_scaled = scaler.transform(X)

# Here we taking 80% for traning and 20% for Testing 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=30)

class_weights = {'SIRA':10, 'BARBUNYA':1, 'BOMBAY':1, 'CALI':1, 'DERMASON':1, 'HOROZ':1, 'SEKER':1}
dt = DecisionTreeClassifier(class_weight=class_weights,max_depth=25,random_state=42)

smote = SMOTE(random_state=42)
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

dt.fit(X_train_resampled, y_train_resampled)
y_pred = dt.predict(X_test)
smdtacc=accuracy_score(y_test, y_pred)
print('SMOTE Decision Tree Accuracy with changing the class weights:\n')

print("Accuracy: ",smdtacc*100,"\n")
classification_rep = classification_report(y_test, y_pred)
print(classification_rep)

print('------------------------------------------------------')
print("\nBefore Descision Tree Results:")
print("Accuracy : ",dt_accuracy*100)
print(reportdt)

SMOTE Decision Tree Accuracy with changing the class weights:

Accuracy:  91.36981270657363 

              precision    recall  f1-score   support

    BARBUNYA       0.92      0.92      0.92       263
      BOMBAY       1.00      1.00      1.00        99
        CALI       0.92      0.91      0.92       297
    DERMASON       0.93      0.88      0.90       705
       HOROZ       0.94      0.96      0.95       408
       SEKER       0.93      0.94      0.93       407
        SIRA       0.84      0.89      0.87       544

    accuracy                           0.91      2723
   macro avg       0.93      0.93      0.93      2723
weighted avg       0.91      0.91      0.91      2723

------------------------------------------------------

Before Descision Tree Results:
Accuracy :  89.46015424164524
              precision    recall  f1-score   support

    BARBUNYA       0.87      0.91      0.89       263
      BOMBAY       1.00      1.00      1.00        99
        CALI       0.92      

### Method 2 : Using GRID(Hyperparameter Tuning) and SMOTE for Descision Tree 

In [11]:
import pandas as pd
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

param_grid = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

clf = DecisionTreeClassifier()

grid_search = GridSearchCV(estimator=clf, param_grid=param_grid, scoring='f1_weighted', n_jobs=-1)

smote = SMOTE(random_state=25)
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

grid_search.fit(X_train_resampled, y_train_resampled)
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(accuracy*100)

report = classification_report(y_test, y_pred)
print("Classification Report:\n", report)

print('------------------------------------------------------')
print("\nBefore Descision Tree Results:")
print("Accuracy : ",dt_accuracy*100)
print(reportdt)

92.47153874403232
Classification Report:
               precision    recall  f1-score   support

    BARBUNYA       0.91      0.93      0.92       263
      BOMBAY       1.00      1.00      1.00        99
        CALI       0.91      0.93      0.92       297
    DERMASON       0.92      0.92      0.92       705
       HOROZ       0.96      0.94      0.95       408
       SEKER       0.96      0.95      0.95       407
        SIRA       0.88      0.88      0.88       544

    accuracy                           0.92      2723
   macro avg       0.93      0.94      0.93      2723
weighted avg       0.92      0.92      0.92      2723

------------------------------------------------------

Before Descision Tree Results:
Accuracy :  89.46015424164524
              precision    recall  f1-score   support

    BARBUNYA       0.87      0.91      0.89       263
      BOMBAY       1.00      1.00      1.00        99
        CALI       0.92      0.90      0.91       297
    DERMASON       0.88    

###
The above report summarizes the performance of a machine learning model on different classes.

1. Precision:-  
    1. Measure of the model's ability to make precise predictions
    2. How many instances predict certain class are actually the classes in dataset. 
    3. Calculated for each class indivisually.
    4. Above Model provides time about 91% correctness. 
 
2. Recall:-
    1. Sensitivity or true positive rate 
    2. Measures the model's ability to identify all positive instances 
    3. Calculated for each class indivisually.
    4. Above Model provides about 90% actaul instance.

3. F1-Score:- 
    1. Harmonic mean of precision and recall
    2. It's a single value that balances both precision and recall
    3. Calculated for each class indivisually.
    4. Above Model indicating a balance between precision and recall
    
4. Support:- 
    1. Number of instances for each class in the test dataset
    
5. Accuracy:- 
    1. The overall accuracy of the model across all classes. 
    2. Above Model indicating accuracy is 91%
    
6. Macro Avg:-
    1. The macro average takes the average of precision, recall, and F1-score for all classes without considering class imbalance. 
    2. Here macro average is 93%.
    
7. Weighted Avg:-
    1. The weighted average takes the average of precision, recall, and F1-score for all classes, considering class imbalance.
    2. Here Weighted Avg is 91%.


Conclusion:

Method 1: 
Here we seen that In order to handle class imbalance, the Decision Tree moddel with SMOTE performs better than the model without SMOTE, obtaining better overall F1-scores and a higher accuracy (91.36% vs. 89.46%). SMOTE improves the model's classification accuracy while helping in balancing class performance.
When SIRA class not affect the other if it is posinous.


Method 2: 
Using SMOTE and HYPERPARAMETER TUNING on DT, it improves the performance of the model with higher accuracy and F1 score. Which is used to handle the imbalanced data over here. We can see through the previous results of DT and after appliying this method on DT model. 